File: code/examples/node/01_hello_claude.js
// 运行: node 01_hello_claude.js
// 依赖: npm install @anthropic-ai/sdk
//
// 这个文件演示了与 Claude API 交互的最基础用法。
// 可以把它理解成：
//   就像 fetch('/api/chat', { method: 'POST', body: JSON.stringify({messages}) })
//   只是 Anthropic SDK 帮你封装了细节，让你专注写对话逻辑。

import Anthropic from '@anthropic-ai/sdk';

// ─────────────────────────────────────────────
// 初始化客户端
// SDK 会自动读取环境变量 ANTHROPIC_API_KEY
// 也可以手动传入：new Anthropic({ apiKey: '...' })
// ─────────────────────────────────────────────
const client = new Anthropic();

// ═══════════════════════════════════════════════════════════════
// 第一节：最简单的一次对话（单轮）
// ═══════════════════════════════════════════════════════════════

/**
 * 最基础的 API 调用示例
 * 对应 HTTP POST /v1/messages，SDK 帮你处理 headers / JSON 序列化
 *
 * @returns {Promise<void>}
 */
async function basicChat() {
  console.log('=== 1. 基础单轮对话 ===\n');

  // messages.create 是核心方法，类似 fetch 的作用
  const response = await client.messages.create({
    model: 'claude-opus-4-5',          // 使用哪个模型
    max_tokens: 1024,                   // 最多生成多少 token（大约 750 个英文单词）
    messages: [
      {
        role: 'user',                   // 角色：user = 你说的话
        content: '用一句话解释什么是 API？',
      },
    ],
  });

  // ─── 解析 response 结构 ────────────────────────────────────
  // response.id          - 本次请求的唯一 ID，方便排查问题
  // response.type        - 固定是 'message'
  // response.role        - 固定是 'assistant'
  // response.model       - 实际使用的模型名称
  // response.stop_reason - 为什么停止：'end_turn' / 'max_tokens' / 'tool_use'
  // response.usage       - token 用量统计
  //   .input_tokens      - 你发送的 token 数
  //   .output_tokens     - Claude 回复的 token 数
  // response.content     - 数组，每个元素是一个内容块
  //   [{ type: 'text', text: '...' }]

  console.log('Claude 说：', response.content[0].text);
  console.log('\n📊 Token 用量：');
  console.log('  输入：', response.usage.input_tokens, 'tokens');
  console.log('  输出：', response.usage.output_tokens, 'tokens');
  console.log('  停止原因：', response.stop_reason);
}

// ═══════════════════════════════════════════════════════════════
// 第二节：System Prompt（给 AI 设定角色 / 行为规则）
// ═══════════════════════════════════════════════════════════════

/**
 * System prompt 就像"员工手册"——告诉 Claude 它是谁、怎么工作。
 * 它在整个对话过程中始终有效，用户看不到也改不掉。
 *
 * @returns {Promise<void>}
 */
async function withSystemPrompt() {
  console.log('\n=== 2. 使用 System Prompt ===\n');

  const response = await client.messages.create({
    model: 'claude-opus-4-5',
    max_tokens: 512,
    // system 字段：角色设定 / 行为规范 / 输出格式要求
    system: `你是一位专门为编程初学者服务的老师。
规则：
1. 回答必须使用中文
2. 用最简单的比喻解释概念
3. 每个解释控制在 3 句话以内`,
    messages: [
      { role: 'user', content: '什么是函数？' },
    ],
  });

  console.log('Claude（扮演老师）：', response.content[0].text);
}

// ═══════════════════════════════════════════════════════════════
// 第三节：多轮对话（对话历史 = messages 数组）
// ═══════════════════════════════════════════════════════════════

/**
 * Claude 本身是"无记忆"的——每次请求都是独立的。
 * 要让它"记住"上下文，你需要把历史消息一起发过去。
 * 就像你跟朋友聊天，发新消息时把整段聊天记录都截图发过去一样。
 *
 * @returns {Promise<void>}
 */
async function multiTurnChat() {
  console.log('\n=== 3. 多轮对话 ===\n');

  // 对话历史：你来我往的消息数组
  // 规则：user 和 assistant 必须交替出现，且必须以 user 开头
  const conversationHistory = [];

  /**
   * 发送一条消息并把 AI 回复追加到历史中
   * @param {string} userMessage - 用户输入
   * @returns {Promise<string>} - AI 的回复文本
   */
  async function chat(userMessage) {
    // 把用户消息加入历史
    conversationHistory.push({
      role: 'user',
      content: userMessage,
    });

    const response = await client.messages.create({
      model: 'claude-opus-4-5',
      max_tokens: 512,
      messages: conversationHistory, // 把完整历史都发过去
    });

    const assistantReply = response.content[0].text;

    // 把 AI 回复也加入历史，下一轮请求时一并发送
    conversationHistory.push({
      role: 'assistant',
      content: assistantReply,
    });

    return assistantReply;
  }

  // 模拟三轮对话
  const reply1 = await chat('我叫小明，我在学 JavaScript。');
  console.log('第1轮 User: 我叫小明，我在学 JavaScript。');
  console.log('第1轮 Claude:', reply1);

  const reply2 = await chat('你还记得我叫什么吗？我在学什么？');
  console.log('\n第2轮 User: 你还记得我叫什么吗？我在学什么？');
  console.log('第2轮 Claude:', reply2);

  const reply3 = await chat('给我推荐一个适合我当前水平的学习资源。');
  console.log('\n第3轮 User: 给我推荐一个适合我当前水平的学习资源。');
  console.log('第3轮 Claude:', reply3);

  console.log('\n当前对话历史长度：', conversationHistory.length, '条消息');
}

// ═══════════════════════════════════════════════════════════════
// 第四节：流式输出（Stream）
// ═══════════════════════════════════════════════════════════════

/**
 * 普通模式：等 Claude 写完整段话再一次性返回（等待时屏幕空白）
 * 流式模式：Claude 写一个字，你就立刻看到一个字（像打字机一样）
 *
 * 适合：聊天界面、长文本生成，能大大改善用户体验。
 *
 * @returns {Promise<void>}
 */
async function streamingChat() {
  console.log('\n=== 4. 流式输出（Stream）===\n');
  console.log('Claude 正在流式回答：\n');

  // 使用 stream() 方法代替 create()
  const stream = client.messages.stream({
    model: 'claude-opus-4-5',
    max_tokens: 256,
    messages: [
      { role: 'user', content: '用三点描述 Node.js 的优势，简短一些。' },
    ],
  });

  // for await...of 逐个处理事件
  // 这就像读水管里的水：水来一点处理一点，不用等水桶装满
  for await (const event of stream) {
    // 只关心文本增量事件
    if (
      event.type === 'content_block_delta' &&
      event.delta.type === 'text_delta'
    ) {
      // process.stdout.write 不换行，实现打字机效果
      process.stdout.write(event.delta.text);
    }
  }

  // 流结束后获取完整 response（包含 usage 等信息）
  const finalMessage = await stream.getFinalMessage();
  console.log('\n\n流式结束，总 token：',
    finalMessage.usage.input_tokens + finalMessage.usage.output_tokens
  );
}

// ═══════════════════════════════════════════════════════════════
// 第五节：完整 response 结构解析（放大镜看 response）
// ═══════════════════════════════════════════════════════════════

/**
 * 打印 response 对象的全部字段，帮助理解数据结构
 *
 * @returns {Promise<void>}
 */
async function inspectResponse() {
  console.log('\n=== 5. Response 结构解析 ===\n');

  const response = await client.messages.create({
    model: 'claude-opus-4-5',
    max_tokens: 64,
    messages: [{ role: 'user', content: 'Hi' }],
  });

  console.log('完整 response 字段：');
  console.log({
    id: response.id,                  // 请求 ID，格式如 msg_01Xk...
    type: response.type,              // 永远是 'message'
    role: response.role,              // 永远是 'assistant'
    model: response.model,            // 实际调用的模型
    stop_reason: response.stop_reason,// 停止原因
    stop_sequence: response.stop_sequence, // 触发停止的特殊序列（通常为 null）
    usage: response.usage,            // { input_tokens, output_tokens }
    content_length: response.content.length, // content 数组长度
    first_content_type: response.content[0]?.type, // 通常是 'text'
  });
}

// ═══════════════════════════════════════════════════════════════
// 主函数：依次运行所有示例
// ═══════════════════════════════════════════════════════════════

async function main() {
  console.log('🚀 Claude API 基础示例\n');
  console.log('类比：这就像 fetch(\'/api/chat\', { method: \'POST\', body: ... })');
  console.log('SDK 帮你处理了鉴权、重试、JSON 解析等脏活\n');
  console.log('─'.repeat(50));

  try {
    await basicChat();
    await withSystemPrompt();
    await multiTurnChat();
    await streamingChat();
    await inspectResponse();

    console.log('\n✅ 所有示例运行完毕！');
    console.log('\n💡 下一步：查看 02_tool_use_basic.js 学习工具调用');
  } catch (error) {
    // 常见错误处理
    if (error.status === 401) {
      console.error('❌ 认证失败：请检查 ANTHROPIC_API_KEY 环境变量');
    } else if (error.status === 429) {
      console.error('❌ 请求过于频繁：稍等几秒后重试');
    } else if (error.status === 529) {
      console.error('❌ API 服务过载：稍后重试');
    } else {
      console.error('❌ 发生错误：', error.message);
    }
    process.exit(1);
  }
}

main();

File: code/examples/python/01_hello_claude.py
# 运行: python 01_hello_claude.py
# 依赖: pip install anthropic
# 说明: 这是学习 Claude API 的第一个文件，从最基础的调用开始
#       类比：就像你第一次学打电话——先学怎么拨号，再学怎么说话

import anthropic  # 导入 Anthropic 官方 SDK，类似 JS 里 import fetch
import asyncio    # Python 异步编程库，用来写 async/await 代码

# ============================================================
# 第一部分：最基础的同步调用（最简单的打招呼方式）
# ============================================================

def basic_hello():
    """
    最简单的 Claude 调用示例
    类比：就像发一条短信，等对方回复，再看回复内容
    """
    print("=" * 50)
    print("【示例1】基础文本调用")
    print("=" * 50)

    # 创建客户端，SDK 会自动读取环境变量 ANTHROPIC_API_KEY
    # 类比：就像打开微信，自动登录你的账号
    client = anthropic.Anthropic()

    # 发送一条消息给 Claude，等待完整回复
    # client.messages.create 是核心方法，类似 JS 的 fetch("/api/chat", {method:"POST"})
    message = client.messages.create(
        model="claude-opus-4-8",          # 指定使用哪个 Claude 模型（型号）
        max_tokens=1024,                   # 最多生成多少个 token（约等于字数上限）
        messages=[                         # 对话历史列表，每条消息是一个字典
            {
                "role": "user",            # role 是发言角色，"user" 表示用户在说话
                "content": "你好，请用一句话介绍你自己。"  # content 是具体说的话
            }
        ]
    )

    # 解析返回的 response 对象
    # message.content 是一个列表，每个元素是一个内容块（通常只有一个文本块）
    reply_text = message.content[0].text  # 取第一个内容块的 .text 属性

    print(f"Claude 回复：{reply_text}")
    print()

    # 解析更多响应字段——response 对象包含很多有用的信息
    print("【响应对象详解】")
    print(f"  stop_reason    = {message.stop_reason}")
    # stop_reason 表示 Claude 为什么停止生成
    # "end_turn" = 正常结束（把话说完了）
    # "max_tokens" = 达到 token 上限被截断了
    # "tool_use" = 要调用工具（见文件2）
    # "stop_sequence" = 遇到了预设的停止符

    print(f"  model          = {message.model}")
    # 实际使用的模型名称（有时和你传入的略有差异）

    print(f"  input_tokens   = {message.usage.input_tokens}")
    # 你发送的内容消耗了多少 token（token 类比：汉字约 1.5-2 个 token）

    print(f"  output_tokens  = {message.usage.output_tokens}")
    # Claude 回复消耗了多少 token

    print(f"  total_tokens   = {message.usage.input_tokens + message.usage.output_tokens}")
    # 本次对话总消耗（计费依据）

    print(f"  message.id     = {message.id}")
    # 每条消息都有唯一 ID，方便追踪和调试

    print()
    return reply_text


# ============================================================
# 第二部分：流式输出（streaming）
# ============================================================

def streaming_hello():
    """
    流式输出示例：Claude 边生成边返回，不用等全部生成完
    类比：就像看弹幕直播，字一个个出来，而不是等主播说完再一次性显示
    适合场景：回复很长时，用户不用干等，体验更好
    """
    print("=" * 50)
    print("【示例2】流式输出 Streaming")
    print("=" * 50)

    client = anthropic.Anthropic()

    print("Claude 正在回复（流式）：", end="", flush=True)

    # 使用 client.messages.stream() 上下文管理器开启流式模式
    # 类比：就像订阅一个直播频道，数据包来一个处理一个
    with client.messages.stream(
        model="claude-opus-4-8",
        max_tokens=300,
        messages=[
            {
                "role": "user",
                "content": "请用3句话描述今天的天气可能是什么样的，要有想象力。"
            }
        ]
    ) as stream:
        # stream.text_stream 是一个生成器，每次 yield 一小段文字
        # 类比：水龙头打开，水一滴滴流出来，我们接一滴处理一滴
        for text_chunk in stream.text_stream:
            print(text_chunk, end="", flush=True)
            # end="" 表示不换行，flush=True 表示立即刷新到屏幕
            # 这样就能看到文字一个个出现的效果

    print("\n")  # 流结束后换行

    # 流结束后，可以获取完整的最终消息对象
    final_message = stream.get_final_message()
    print(f"  流式完成，总 tokens: {final_message.usage.input_tokens + final_message.usage.output_tokens}")
    print()


# ============================================================
# 第三部分：System Prompt（给 Claude 设定角色和规则）
# ============================================================

def system_prompt_demo():
    """
    System Prompt 示例：在对话开始前给 Claude 下"规矩"
    类比：就像你雇了一个新员工，先给他写一份岗位说明书
          "你是客服人员，只能回答产品相关问题，用礼貌语气"
    """
    print("=" * 50)
    print("【示例3】System Prompt 使用")
    print("=" * 50)

    client = anthropic.Anthropic()

    # system 参数放在 messages 列表之外，单独传入
    # 它会作为整个对话的"背景设定"，用户看不到，但 Claude 始终遵守
    message = client.messages.create(
        model="claude-opus-4-8",
        max_tokens=200,
        system=(
            "你是一位古代诗人，名叫李白。"
            "你只能用古诗词的风格回答问题。"
            "每句话必须押韵，不超过四行。"
        ),
        messages=[
            {
                "role": "user",
                "content": "今天天气真好，你心情怎么样？"
            }
        ]
    )

    print(f"（角色：李白）Claude 回答：\n{message.content[0].text}")
    print()


# ============================================================
# 第四部分：多轮对话（维护对话历史）
# ============================================================

def multi_turn_conversation():
    """
    多轮对话示例：Claude 没有自动记忆，每次都要把历史带上
    类比：就像给客服发邮件，每次都要 CC 之前的邮件链，
          对方才知道你们之前聊了什么
    """
    print("=" * 50)
    print("【示例4】多轮对话")
    print("=" * 50)

    client = anthropic.Anthropic()

    # messages 列表就是整个对话历史
    # 每次调用后，把 Claude 的回复也加进去，下次一起发送
    messages = []  # 从空对话开始

    # --- 第一轮 ---
    user_input_1 = "我叫小明，我喜欢吃饺子。"
    print(f"用户：{user_input_1}")

    messages.append({"role": "user", "content": user_input_1})  # 加入用户消息

    response_1 = client.messages.create(
        model="claude-opus-4-8",
        max_tokens=100,
        messages=messages  # 发送完整历史（目前只有1条）
    )

    assistant_reply_1 = response_1.content[0].text
    print(f"Claude：{assistant_reply_1}")

    # 把 Claude 的回复也加入历史，下一轮就能"记住"了
    messages.append({"role": "assistant", "content": assistant_reply_1})

    # --- 第二轮 ---
    user_input_2 = "你还记得我叫什么名字吗？我喜欢吃什么？"
    print(f"\n用户：{user_input_2}")

    messages.append({"role": "user", "content": user_input_2})

    response_2 = client.messages.create(
        model="claude-opus-4-8",
        max_tokens=100,
        messages=messages  # 发送包含第一轮在内的完整历史（共3条）
    )

    assistant_reply_2 = response_2.content[0].text
    print(f"Claude：{assistant_reply_2}")

    # 打印当前对话历史长度
    messages.append({"role": "assistant", "content": assistant_reply_2})
    print(f"\n  当前对话历史共 {len(messages)} 条消息")
    print()


# ============================================================
# 第五部分：async 异步版本（与同步版本对比）
# ============================================================

async def async_hello():
    """
    async 异步版本：适合在 Web 服务器或同时处理多个请求的场景
    类比：同步版 = 去银行排队，一个人办完才轮到你
          异步版 = 拿号等叫号，等待时可以去别处办其他事
    什么时候用 async：
      - 你的程序同时要处理很多用户请求（Web API）
      - 你想同时调用多个 AI 接口不互相等待
    什么时候用同步：
      - 简单脚本、一次性任务、初学练习
    """
    print("=" * 50)
    print("【示例5】async 异步版本")
    print("=" * 50)

    # AsyncAnthropic 是异步版客户端，其他参数完全一样
    # 类比：AsyncAnthropic 是"异步模式银行"，Anthropic 是"普通排队银行"
    client = anthropic.AsyncAnthropic()

    # await 表示"在这里等一下，但不阻塞其他任务"
    # 类比：await = "等我外卖到了通知我，我先去做别的事"
    message = await client.messages.create(
        model="claude-opus-4-8",
        max_tokens=100,
        messages=[
            {
                "role": "user",
                "content": "用一句话说明同步和异步的区别。"
            }
        ]
    )

    print(f"（异步）Claude 回复：{message.content[0].text}")
    print()


async def async_streaming_hello():
    """
    async 流式输出：异步 + 流式的组合，适合生产环境 Web 服务
    """
    print("=" * 50)
    print("【示例6】async 流式输出")
    print("=" * 50)

    client = anthropic.AsyncAnthropic()

    print("（异步流式）Claude 回复：", end="", flush=True)

    # async with 是异步版的 with，async for 是异步版的 for
    async with client.messages.stream(
        model="claude-opus-4-8",
        max_tokens=150,
        messages=[
            {
                "role": "user",
                "content": "用两句话介绍 Python。"
            }
        ]
    ) as stream:
        # async for 会异步等待每个 chunk 到来
        async for text_chunk in stream.text_stream:
            print(text_chunk, end="", flush=True)

    print("\n")


# ============================================================
# 主程序入口
# ============================================================

def main():
    """
    主函数：依次运行所有示例
    """
    print("\n🚀 Claude API 基础示例开始运行\n")

    # 示例1：基础调用
    basic_hello()

    # 示例2：流式输出
    streaming_hello()

    # 示例3：System Prompt
    system_prompt_demo()

    # 示例4：多轮对话
    multi_turn_conversation()

    # 示例5 & 6：async 版本（需要用 asyncio.run 来运行 async 函数）
    # asyncio.run() 是启动 async 程序的"开关"，只能在最外层调用一次
    asyncio.run(async_hello())
    asyncio.run(async_streaming_hello())

    print("✅ 所有示例运行完毕！")
    print()
    print("学到了什么？")
    print("  1. client.messages.create() = 基础调用，等待完整回复")
    print("  2. client.messages.stream() = 流式调用，边生成边输出")
    print("  3. system= 参数 = 给 Claude 设定角色和规则")
    print("  4. messages 列表 = 手动维护对话历史（Claude 没有自动记忆）")
    print("  5. AsyncAnthropic = 异步版，适合并发场景")


if __name__ == "__main__":
    # 这是 Python 标准入口写法
    # 意思是：只有直接运行这个文件时才执行 main()
    # 如果被其他文件 import，则不自动执行
    main()
